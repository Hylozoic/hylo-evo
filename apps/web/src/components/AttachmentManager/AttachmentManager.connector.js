import { connect } from 'react-redux'
import {
  setAttachments,
  clearAttachments,
  addAttachment,
  removeAttachment,
  moveAttachment,
  getAttachments,
  getAttachmentsFromObject,
  getUploadAttachmentPending
} from './AttachmentManager.store'

export function mapStateToProps (state, props) {
  return {
    uploadAttachmentPending: getUploadAttachmentPending(state, props),
    attachments: getAttachments(state, props),
    attachmentsFromObject: getAttachmentsFromObject(state, props)
  }
}

export const mapDispatchToProps = {
  addAttachment,
  removeAttachment,
  moveAttachment,
  setAttachments,
  clearAttachments
}

export const mergeProps = (stateProps, dispatchProps, ownProps) => {
  const { type, id, attachmentType } = ownProps
  const { attachmentsFromObject } = stateProps
  const { moveAttachment, setAttachments, clearAttachments } = dispatchProps
  return {
    ...stateProps,
    ...dispatchProps,
    ...ownProps,
    switchImages: (position1, position2) => moveAttachment(type, id, 'image', position1, position2),
    loadAttachments: () => setAttachments(type, id, attachmentType, attachmentsFromObject),
    clearAttachments: () => clearAttachments(type, id, attachmentType)
  }
}

export default connect(mapStateToProps, mapDispatchToProps, mergeProps, { forwardRef: true })
