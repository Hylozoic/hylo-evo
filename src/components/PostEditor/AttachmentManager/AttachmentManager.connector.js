import { get, isEmpty } from 'lodash/fp'
import { connect } from 'react-redux'
import { UPLOAD_ATTACHMENT } from 'store/constants'
import {
  addAttachment,
  removeAttachment,
  switchAttachments,
  setAttachments,
  getAttachments,
  makeAttachmentSelector
} from './AttachmentManager.store'

export function mapStateToProps (state, props) {
  const pending = get(['pending', UPLOAD_ATTACHMENT, 'attachmentType'], state) === props.type
  const attachments = getAttachments(state, props)
  const attachmentsFromPost = makeAttachmentSelector(props.type)(state, props)
  const showAttachments = !isEmpty(attachments) || pending  // last clause is for testing only

  return {
    attachments,
    attachmentsFromPost,
    showAttachments,
    pending
  }
}

export const mapDispatchToProps = {
  addAttachment,
  removeAttachment,
  switchAttachments,
  setAttachments
}

export const mergeProps = (stateProps, dispatchProps, ownProps) => {
  const { type } = ownProps
  const { attachmentsFromPost } = stateProps
  const { addAttachment, removeAttachment, switchAttachments, setAttachments } = dispatchProps

  return {
    ...stateProps,
    ...dispatchProps,
    ...ownProps,
    addAttachment: url => addAttachment(url, type),
    removeAttachment: position => removeAttachment(position, type),
    switchAttachments: (position1, position2) => switchAttachments(position1, position2, type),
    loadAttachments: () => setAttachments(attachmentsFromPost.map(attachment => attachment.url), type),
    clearAttachments: () => setAttachments([], type)
  }
}

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)
