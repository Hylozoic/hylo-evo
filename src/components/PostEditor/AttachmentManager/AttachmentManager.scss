// ImagePreviews

.section-label {
  composes: section-label from '../PostEditor.scss';
}

.image-preview-section {
  display: flex;
  flex-direction: row;
  align-items: center;
  padding: 15px 20px 0px;
  border-top: solid 1px $color-ghost-30;
}

.image-previews {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  align-items: center;
}

.image-box {
  width: 100px;
  height: 100px;
  border-radius: 4px;
  margin-right: 15px;
  margin-bottom: 15px;
}

.image-preview {
  composes: image-box;
}

.image {
  composes: image-box;
  background-size: cover;
  background-position: center;
  position: relative;
  padding: 8px;
}

.remove-button {
  position: absolute;
  top: -7px;
  right: -7px;
  border-radius: 50%;
  color: white;
  background-color: $color-rhino;
  font-size: 12px;
  padding: 1px;
  border: 1px solid white;
  background-clip: padding-box;
  cursor: pointer;
}

.drag-preview {
  width: 84px;
  height: 84px;
}

.add-image {
  composes: image-box;
  padding-top: 16px;
  text-align: center;
  border: 2px dashed $color-loblolly;
  font-size: 40px;
  color: $color-rhino-30;
  cursor: pointer;
}
