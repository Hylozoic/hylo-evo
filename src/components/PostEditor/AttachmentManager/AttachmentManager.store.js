import { pullAt, clone, get } from 'lodash/fp'
import { createSelector as ormCreateSelector } from 'redux-orm'
import orm from 'store/models'

export const MODULE_NAME = 'AttachmentManager'
export const LOAD_ATTACHMENTS = `${MODULE_NAME}/LOAD_ATTACHMENTS`
export const SET_ATTACHMENTS = `${MODULE_NAME}/SET_ATTACHMENTS`
export const ADD_ATTACHMENT = `${MODULE_NAME}/ADD_ATTACHMENT`
export const REMOVE_ATTACHMENT = `${MODULE_NAME}/REMOVE_ATTACHMENT`
export const SWITCH_ATTACHMENTS = `${MODULE_NAME}/SWITCH_ATTACHMENTS`

// action generators

export function setAttachments (attachments, type) {
  return {
    type: SET_ATTACHMENTS,
    payload: {
      attachments,
      type
    }
  }
}

export function addAttachment (url, type) {
  return {
    type: ADD_ATTACHMENT,
    payload: {
      url,
      type
    }
  }
}

export function removeAttachment (position, type) {
  return {
    type: REMOVE_ATTACHMENT,
    payload: {
      position,
      type
    }
  }
}

export function switchAttachments (position1, position2, type) {
  return {
    type: SWITCH_ATTACHMENTS,
    payload: {
      position1,
      position2,
      type
    }
  }
}

// Selectors

export function getAttachments (state, { type }) {
  return state[MODULE_NAME][type]
}

export const makeAttachmentSelector = attachmentType => ormCreateSelector(
  orm,
  get('orm'),
  (state, { postId }) => postId,
  ({ Attachment }, postId) =>
    Attachment.all()
    .filter(({ type, post }) =>
      type === attachmentType && post === postId)
    .orderBy('position')
    .toModelArray()
  )

// Reducer

export const defaultState = {
  image: [],
  file: []
}

export default function reducer (state = defaultState, action) {
  const { error, type, payload } = action
  if (error) return state

  switch (type) {
    case SET_ATTACHMENTS:
      return {...state, [payload.type]: payload.attachments}
    case ADD_ATTACHMENT:
      return {...state, [payload.type]: state[payload.type].concat([payload.url])}
    case REMOVE_ATTACHMENT:
      return {...state, [payload.type]: pullAt(payload.position, state[payload.type])}
    case SWITCH_ATTACHMENTS:
      const { position1, position2 } = payload
      const attachments = clone(state[payload.type])
      attachments[position1] = state[payload.type][position2]
      attachments[position2] = state[payload.type][position1]
      return {
        ...state,
        [payload.type]: attachments
      }
    default:
      return state
  }
}
