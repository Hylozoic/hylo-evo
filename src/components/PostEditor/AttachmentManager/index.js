import component from './AttachmentManager'
import connector from './AttachmentManager.connector'

export default connector(component)
