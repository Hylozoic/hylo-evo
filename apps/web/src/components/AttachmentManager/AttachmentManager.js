import { isEmpty, filter } from 'lodash/fp'
import path from 'path'
import PropTypes from 'prop-types'
import React from 'react'
import { useTranslation } from 'react-i18next'
import { DndContext } from '@dnd-kit/core'
import { SortableContext, horizontalListSortingStrategy, useSortable } from '@dnd-kit/sortable'
import { CSS } from '@dnd-kit/utilities'
import Icon from 'components/Icon'
import Loading from 'components/Loading'
import UploadAttachmentButton from 'components/UploadAttachmentButton'
import { bgImageStyle } from 'util/index'
import { ID_FOR_NEW } from './AttachmentManager.store'
import classes from './AttachmentManager.module.scss'

export const attachmentsObjectType = {
  id: PropTypes.string,
  url: PropTypes.string.isRequired,
  attachmentType: PropTypes.string.isRequired
}

export const attachmentsFromObjectType = {
  type: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  ...attachmentsObjectType
}

export default class AttachmentManager extends React.Component {
  static propTypes = {
    type: PropTypes.string.isRequired,
    id: PropTypes.string,
    attachmentType: PropTypes.string,
    showAddButton: PropTypes.bool,
    showLabel: PropTypes.bool,
    showLoading: PropTypes.bool,
    // provided by connector
    uploadAttachmentPending: PropTypes.bool,
    attachments: PropTypes.arrayOf(PropTypes.shape(attachmentsObjectType)),
    attachmentsFromObject: PropTypes.arrayOf(PropTypes.shape(attachmentsFromObjectType)),
    loadAttachments: PropTypes.func.isRequired,
    addAttachment: PropTypes.func.isRequired,
    removeAttachment: PropTypes.func.isRequired,
    moveAttachment: PropTypes.func.isRequired,
    clearAttachments: PropTypes.func.isRequired,
    // only used by connector
    setAttachments: PropTypes.func.isRequired
  }

  static defaultProps = {
    id: ID_FOR_NEW
  }

  componentDidMount () {
    this.props.loadAttachments()
  }

  componentDidUpdate (prevProps) {
    if (
      isEmpty(prevProps.attachmentsFromObject) &&
      !isEmpty(this.props.attachmentsFromObject)
    ) {
      this.props.loadAttachments()
    }
  }

  componentWillUnmount () {
    this.props.clearAttachments()
  }

  render () {
    const { attachments, attachmentType, uploadAttachmentPending, showLoading } = this.props

    if (isEmpty(attachments) && !uploadAttachmentPending) return null

    // TODO: put this in a state variable that can change in real time to update the ordering as you drag
    const imageAttachments = filter({ attachmentType: 'image' }, attachments)
    const fileAttachments = filter({ attachmentType: 'file' }, attachments)
    const showImages = (!isEmpty(imageAttachments) || (uploadAttachmentPending && showLoading)) &&
      (!attachmentType || attachmentType === 'image')
    const showFiles = (!isEmpty(fileAttachments) || (uploadAttachmentPending && showLoading)) &&
      (!attachmentType || attachmentType === 'file')

    return (
      <>
        {showImages &&
          <ImageManager {...this.props} showLoading={showLoading} attachments={imageAttachments} />}
        {showFiles &&
          <FileManager {...this.props} showLoading={showLoading} attachments={fileAttachments} />}
      </>
    )
  }
}

export function ImageManager (props) {
  const { t } = useTranslation()
  const {
    type, id, attachments, addAttachment, removeAttachment,
    uploadAttachmentPending, showLoading, showAddButton, showLabel
  } = props

  const handleDragEnd = ({ active, over }) => {
    if (active.id !== over.id) {
      props.switchImages(active.data.current.sortable.index, over.data.current.sortable.index)
    }
  }

  const images = attachments.map((attachment, i) => ({
    ...attachment,
    id: attachment.url
  }))

  return (
    <DndContext onDragEnd={handleDragEnd}>
      <SortableContext items={images} strategy={horizontalListSortingStrategy}>
        <div className={classes.imageManager}>
          {showLabel && <div className={classes.sectionLabel}>{t('Images')}</div>}
          <div className={classes.imagePreviews}>
            {images.map((attachment, i) =>
              <ImagePreview
                attachment={attachment}
                removeImage={() => removeAttachment(type, id, attachment)}
                index={i}
                key={i}
              />)}
            {showLoading && uploadAttachmentPending && <div className={classes.addImage}><Loading /></div>}
            {showAddButton && (
              <UploadAttachmentButton
                type={type}
                id={id}
                attachmentType='image'
                onSuccess={attachment => addAttachment(type, id, attachment)}>
                <div className={classes.addImage}>+</div>
              </UploadAttachmentButton>)}
          </div>
        </div>
      </SortableContext>
    </DndContext>
  )
}

export function ImagePreview (props) {
  const {
    attachment, removeImage
  } = props

  const {
    attributes,
    listeners,
    setNodeRef,
    transform,
    transition
  } = useSortable({
    id: attachment.id,
    transition: null
    // transition: null,{
    //   duration: 150, // milliseconds
    //   easing: 'cubic-bezier(0.25, 1, 0.5, 1)'
    // }
  })

  const style = {
    transform: CSS.Transform.toString(transform),
    transition
  }

  return (
    <div className={classes.imagePreview} ref={setNodeRef} style={style}>
      <Icon name='Ex' className={classes.removeImage} onClick={removeImage} />
      <div style={bgImageStyle(attachment.url)} className={classes.image} {...listeners} {...attributes} />
    </div>
  )
}

export function FileManager ({
  type, id, attachments, addAttachment, removeAttachment,
  uploadAttachmentPending, showLoading, showAddButton, showLabel
}) {
  const { t } = useTranslation()

  return (
    <div className={classes.fileManager}>
      {showLabel && <div className={classes.sectionLabel}>{t('Files')}</div>}
      <div className={classes.filePreviews}>
        {attachments.map((attachment, i) =>
          <FilePreview
            attachment={attachment}
            removeFile={() => removeAttachment(type, id, attachment)}
            key={i} />)}
        {showLoading && uploadAttachmentPending && <div className={classes.loadingFile}>{t('Loading...')}</div>}
        {showAddButton && (
          <UploadAttachmentButton
            id={id}
            type={type}
            attachmentType='file'
            onSuccess={attachment => addAttachment(type, id, attachment)}
            className={classes.addFileRow}>
            <div className={classes.addFile}>
              <span className={classes.addFilePlus}>+</span> {t('Add File')}
            </div>
          </UploadAttachmentButton>)}
      </div>
    </div>
  )
}

export function FilePreview ({ attachment, removeFile, fileSize }) {
  const filename = path.basename(attachment.url)
  return (
    <div className={classes.filePreview}>
      <Icon name='Document' className={classes.iconDocument} />
      <div className={classes.fileName}>{decodeURIComponent(filename)}</div>
      {fileSize && <div className={classes.fileSize}>{fileSize}</div>}
      <Icon name='Ex' className={classes.removeFile} onClick={removeFile} />
    </div>
  )
}
