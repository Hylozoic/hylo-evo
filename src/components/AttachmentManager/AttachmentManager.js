import React, { useEffect, useRef } from 'react'
import { useTranslation } from 'react-i18next'
import { useDrag, useDrop } from 'react-dnd'
import { isEmpty, filter } from 'lodash/fp'
import path from 'path'
import PropTypes from 'prop-types'
import Icon from 'components/Icon'
import Loading from 'components/Loading'
import UploadAttachmentButton from 'components/UploadAttachmentButton'
import { bgImageStyle } from 'util/index'
import { ID_FOR_NEW } from './AttachmentManager.store'
import './AttachmentManager.scss'

export const attachmentsObjectType = {
  url: PropTypes.string.isRequired,
  attachmentType: PropTypes.string.isRequired
}

export const attachmentsFromObjectType = {
  type: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  ...attachmentsObjectType
}

const AttachmentManager = (props) => {
  useEffect(() => {
    props.loadAttachments()
    // return () => props.clearAttachments()
  }, [])
  // useEffect(() => {
  //   props.loadAttachments()
  // }, [props.attachmentsFromObject])

  const { attachments, attachmentType, uploadAttachmentPending, showLoading } = props

  if (isEmpty(attachments) && !uploadAttachmentPending) return null

  // TODO: put this in a state variable that can change in real time to update the ordering as you drag
  const imageAttachments = filter({ attachmentType: 'image' }, attachments)
  const fileAttachments = filter({ attachmentType: 'file' }, attachments)
  const showImages = (!isEmpty(imageAttachments) || (uploadAttachmentPending && showLoading)) &&
      (!attachmentType || attachmentType === 'image')
  const showFiles = (!isEmpty(fileAttachments) || (uploadAttachmentPending && showLoading)) &&
      (!attachmentType || attachmentType === 'file')

  return <>
    {showImages &&
    <ImageManager {...props} showLoading={showLoading} attachments={imageAttachments} />}
    {showFiles &&
    <FileManager {...props} showLoading={showLoading} attachments={fileAttachments} />}
  </>
}

AttachmentManager.propTypes = {
  type: PropTypes.string.isRequired,
  id: PropTypes.string,
  attachmentType: PropTypes.string,
  showAddButton: PropTypes.bool,
  showLabel: PropTypes.bool,
  showLoading: PropTypes.bool,
  // provided by connector
  uploadAttachmentPending: PropTypes.bool,
  attachments: PropTypes.arrayOf(PropTypes.shape(attachmentsObjectType)),
  attachmentsFromObject: PropTypes.arrayOf(PropTypes.shape(attachmentsFromObjectType)),
  loadAttachments: PropTypes.func.isRequired,
  addAttachment: PropTypes.func.isRequired,
  removeAttachment: PropTypes.func.isRequired,
  switchAttachments: PropTypes.func.isRequired,
  clearAttachments: PropTypes.func.isRequired,
  // only used by connector
  setAttachments: PropTypes.func.isRequired
}

AttachmentManager.defaultProps = {
  id: ID_FOR_NEW
}

export const ImageManager = (props) => {
  const {
    type, id, attachments, addAttachment, removeAttachment,
    uploadAttachmentPending, showLoading, showAddButton, showLabel
  } = props
  const { t } = useTranslation()

  return <div styleName='image-manager'>
    {showLabel && <div styleName='section-label'>{t('AttachmentManager.imagesTitle')}</div>}
    <div styleName='image-previews'>
      {attachments.map((attachment, i) =>
        <ImagePreview
          attachment={attachment}
          removeImage={() => removeAttachment(type, id, attachment)}
          switchImages={props.switchImages}
          index={i}
          key={i} />)}
      {showLoading && uploadAttachmentPending && <div styleName='add-image'><Loading /></div>}
      {showAddButton && <UploadAttachmentButton
        type={type}
        id={id}
        attachmentType='image'
        onSuccess={attachment => addAttachment(type, id, attachment)}>
        <div styleName='add-image'>+</div>
      </UploadAttachmentButton>}
    </div>
  </div>
}

export const ImagePreview = (props) => {
  const {
    attachment, index, removeImage, switchImages
  } = props

  const ref = useRef(null)

  const [{ handlerId }, drop] = useDrop({  // eslint-disable-line
    accept: 'image',
    collect (monitor) {
      return {
        handlerId: monitor.getHandlerId()
      }
    },
    hover (item, monitor) {
      if (!ref.current) {
        return
      }
      const dragIndex = item.index
      const hoverIndex = index

      // Don't replace items with themselves
      if (dragIndex === hoverIndex) {
        return
      }
      // Determine rectangle on screen
      const hoverBoundingRect = ref.current?.getBoundingClientRect()
      // Get vertical middle
      const hoverMiddleX =
        (hoverBoundingRect.right - hoverBoundingRect.left) / 2
      // Determine mouse position
      const clientOffset = monitor.getClientOffset()
      // Get pixels to the left
      const hoverClientX = clientOffset.x - hoverBoundingRect.left
      // Only perform the move when the mouse has crossed half of the items width
      // Dragging right
      if (dragIndex < hoverIndex && hoverClientX < hoverMiddleX) {
        return
      }
      // Dragging upwards
      if (dragIndex > hoverIndex && hoverClientX > hoverMiddleX) {
        return
      }
      // Time to actually perform the action
      // XXX: movePost(dragIndex, hoverIndex)
      switchImages(dragIndex, hoverIndex)

      // Note: we're mutating the monitor item here!
      // Generally it's better to avoid mutations,
      // but it's good here for the sake of performance
      // to avoid expensive index searches.
      item.index = hoverIndex
    }
  })
  const [{ isDragging }, drag] = useDrag({
    type: 'image',
    item: () => {
      return { id: attachment.id, index }
    },
    collect: (monitor) => ({
      isDragging: monitor.isDragging()
    })
  })
  const opacity = isDragging ? 0 : 1
  drag(drop(ref))

  return (
    <div styleName='image-preview' ref={ref} style={{ opacity }}>
      <div style={bgImageStyle(attachment.url)} styleName='image'>
        <Icon name='Ex' styleName='remove-image' onClick={removeImage} />
      </div>
    </div>
  )
}

export const FileManager = ({
  type, id, attachments, addAttachment, removeAttachment,
  uploadAttachmentPending, showLoading, showAddButton, showLabel
}) => {
  const { t } = useTranslation()
  return <div styleName='file-manager'>
    {showLabel && <div styleName='section-label'>{t('AttachmentManager.files')}</div>}
    <div styleName='file-previews'>
      {attachments.map((attachment, i) =>
        <FilePreview
          attachment={attachment}
          removeFile={() => removeAttachment(type, id, attachment)}
          key={i} />)}
      {showLoading && uploadAttachmentPending && <div styleName='loading-file'>{t('AttachmentManager.loadingFile')}</div>}
      {showAddButton && <UploadAttachmentButton
        id={id}
        type={type}
        attachmentType='file'
        onSuccess={attachment => addAttachment(type, id, attachment)}
        styleName='add-file-row'>
        <div styleName='add-file'>
          <span styleName='add-file-plus'>+</span> {t('AttachmentManager.addFile')}</div>
      </UploadAttachmentButton>}
    </div>
  </div>
}

export const FilePreview = ({ attachment, removeFile, fileSize }) => {
  const filename = path.basename(attachment.url)
  return <div styleName='file-preview'>
    <Icon name='Document' styleName='icon-document' />
    <div styleName='file-name'>{decodeURIComponent(filename)}</div>
    {fileSize && <div styleName='file-size'>{fileSize}</div>}
    <Icon name='Ex' styleName='remove-file' onClick={removeFile} />
  </div>
}

export default AttachmentManager
