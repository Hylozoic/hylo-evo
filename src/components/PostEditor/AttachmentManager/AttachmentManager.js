import React from 'react'
import Loading from 'components/Loading'
import Icon from 'components/Icon'
import { bgImageStyle } from 'utils/index'
import { DragDropContext, DragSource, DropTarget } from 'react-dnd'
import HTML5Backend from 'react-dnd-html5-backend'
import ChangeImageButton from 'components/ChangeImageButton'
import './AttachmentManager.scss'
import path from 'path'
import { isEmpty } from 'lodash/fp'

export const uploadSettings = id => ({
  type: 'post',
  id: id || 'new'
})

export default class AttachmentManager extends React.Component {
  componentDidMount () {
    this.props.loadAttachments()
  }

  componentDidUpdate (prevProps) {
    if (isEmpty(prevProps.attachmentsFromPost) && !isEmpty(this.props.attachmentsFromPost)) {
      this.props.loadAttachments()
    }
  }

  componentWillUnmount () {
    this.props.clearAttachments()
  }

  render () {
    if (this.props.type === 'image') {
      return <ImageManager {...this.props} />
    } else if (this.props.type === 'file') {
      return <FileManager {...this.props} />
    }
  }
}

export const ImageManager = DragDropContext(HTML5Backend)(
class ImageManager extends React.Component {
  render () {
    const { postId, showAttachments, attachments, pending, addAttachment, removeAttachment, switchAttachments } = this.props
    if (!showAttachments) return null

    return <div styleName='image-manager'>
      <div styleName='section-label'>Images</div>
      <div styleName='image-previews'>
        {attachments.map((url, i) =>
          <ImagePreview url={url}
            removeImage={removeAttachment}
            switchImages={switchAttachments}
            position={i}
            key={i} />)}
        {pending && <div styleName='add-image'><Loading /></div>}
        <ChangeImageButton update={addAttachment}
          uploadSettings={uploadSettings(postId)}
          attachmentType='image'>
          <div styleName='add-image'>+</div>
        </ChangeImageButton>
      </div>
    </div>
  }
})

const imagePreviewSource = {
  beginDrag (props) {
    return {
      position: props.position
    }
  }
}

const imagePreviewTarget = {
  drop (props, monitor, component) {
    const item = monitor.getItem()
    props.switchImages(props.position, item.position)
  }
}

export const ImagePreview = DropTarget('ImagePreview', imagePreviewTarget, (connect, monitor) => ({
  connectDropTarget: connect.dropTarget()
}))(
DragSource('ImagePreview', imagePreviewSource, (connect, monitor) => ({
  connectDragSource: connect.dragSource(),
  connectDragPreview: connect.dragPreview(),
  isDragging: monitor.isDragging()
}))(
class ImagePreview extends React.Component {
  render () {
    const {
      url, removeImage, connectDragSource, connectDragPreview, connectDropTarget, position
    } = this.props

    return connectDropTarget(connectDragSource(<div styleName='image-preview'>
      <div style={bgImageStyle(url)} styleName='image'>
        <Icon name='Ex' styleName='remove-image' onClick={() => removeImage(position)} />
        {connectDragPreview(<div styleName='drag-preview' />)}
      </div>
    </div>))
  }
}))

export function FileManager ({
  postId, showAttachments, attachments, pending, addAttachment, removeAttachment
}) {
  if (!showAttachments) return null

  return <div styleName='file-manager'>
    <div styleName='section-label'>Files</div>
    <div styleName='file-previews'>
      {attachments.map((url, i) =>
        <FilePreview url={url}
          removeFile={removeAttachment}
          position={i}
          key={i} />)}
      {pending && <div styleName='loading-file'>Loading...</div>}
      <ChangeImageButton
        styleName='add-file-row'
        update={addAttachment}
        uploadSettings={uploadSettings(postId)}
        attachmentType='file'>
        <div styleName='add-file'>
          <span styleName='add-file-plus'>+</span> Add File</div>
      </ChangeImageButton>
    </div>
  </div>
}

export function FilePreview ({ url, position, removeFile, fileSize }) {
  const name = path.basename(url)
  return <div styleName='file-preview'>
    <Icon name='Document' styleName='icon-document' />
    <div styleName='file-name'>{decodeURIComponent(name)}</div>
    {fileSize && <div styleName='file-size'>{fileSize}</div>}
    <Icon name='Ex' styleName='remove-file' onClick={() => removeFile(position)} />
  </div>
}
