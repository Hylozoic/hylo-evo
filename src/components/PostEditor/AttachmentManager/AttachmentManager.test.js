import React from 'react'
import { shallow } from 'enzyme'
import AttachmentManager, { ImageManager, ImagePreview, FileManager, FilePreview } from './AttachmentManager'

describe('AttachmentManager', () => {
  const props = {
    prop1: 'foo',
    prop2: 'bar',
    prop3: 'baz'
  }

  it('renders ImageManager when type is image', () => {
    const wrapper = shallow(<AttachmentManager type='image' {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders FileManager when type is file', () => {
    const wrapper = shallow(<AttachmentManager type='file' {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('ImageManager', () => {
  it('matches last snapshot', () => {
    const props = {
      postId: 1,
      addAttachment: () => {},
      removeAttachment: () => {},
      switchAttachments: () => {},
      showAttachments: true,
      pending: true,
      attachments: ['foo.png', 'bar.jpg']
    }
    const wrapper = shallow(<ImageManager.DecoratedComponent {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('ImagePreview', () => {
  it('matches last snapshot', () => {
    const props = {
      url: 'foo.zng',
      removeImage: () => {},
      position: 1,
      connectDragSource: i => i,
      connectDragPreview: i => i,
      connectDropTarget: i => i
    }
    const wrapper = shallow(<ImagePreview.DecoratedComponent {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('FileManager', () => {
  it('matches last snapshot', () => {
    const props = {
      postId: 1,
      addAttachment: () => {},
      removeAttachment: () => {},
      showAttachments: true,
      pending: true,
      attachments: ['foo.pdf', 'bar.zip']
    }
    const wrapper = shallow(<FileManager {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('FilePreview', () => {
  it('matches last snapshot', () => {
    const props = {
      url: 'foo.pdf',
      removeImage: () => {},
      position: 1,
      fileSize: '23.3mb'
    }
    const wrapper = shallow(<FilePreview {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})
