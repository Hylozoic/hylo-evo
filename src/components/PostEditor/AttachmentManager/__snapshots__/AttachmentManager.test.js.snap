// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttachmentManager renders FileManager when type is file 1`] = `
<FileManager
  loadAttachments={[Function]}
  prop1="foo"
  prop2="bar"
  prop3="baz"
  type="file"
/>
`;

exports[`AttachmentManager renders ImageManager when type is image 1`] = `
<DragDropContext(ImageManager)
  loadAttachments={[Function]}
  prop1="foo"
  prop2="bar"
  prop3="baz"
  type="image"
/>
`;

exports[`FileManager matches last snapshot 1`] = `
<div
  data-stylename="file-manager"
>
  <div
    data-stylename="section-label"
  >
    Files
  </div>
  <div
    data-stylename="file-previews"
  >
    <FilePreview
      key="0"
      position={0}
      removeFile={[Function]}
      url="foo.pdf"
    />
    <FilePreview
      key="1"
      position={1}
      removeFile={[Function]}
      url="bar.zip"
    />
    <div
      data-stylename="loading-file"
    >
      Loading...
    </div>
    <Connect(ChangeImageButton)
      attachmentType="file"
      data-stylename="add-file-row"
      update={[Function]}
      uploadSettings={
        Object {
          "id": 1,
          "type": "post",
        }
      }
    >
      <div
        data-stylename="add-file"
      >
        <span
          data-stylename="add-file-plus"
        >
          +
        </span>
         Add File
      </div>
    </Connect(ChangeImageButton)>
  </div>
</div>
`;

exports[`FilePreview matches last snapshot 1`] = `
<div
  data-stylename="file-preview"
>
  <Icon
    data-stylename="icon-document"
    name="Document"
  />
  <div
    data-stylename="file-name"
  >
    foo.pdf
  </div>
  <div
    data-stylename="file-size"
  >
    23.3mb
  </div>
  <Icon
    data-stylename="remove-file"
    name="Ex"
    onClick={[Function]}
  />
</div>
`;

exports[`ImageManager matches last snapshot 1`] = `
<div
  data-stylename="image-manager"
>
  <div
    data-stylename="section-label"
  >
    Images
  </div>
  <div
    data-stylename="image-previews"
  >
    <DropTarget(DragSource(ImagePreview))
      key="0"
      position={0}
      removeImage={[Function]}
      switchImages={[Function]}
      url="foo.png"
    />
    <DropTarget(DragSource(ImagePreview))
      key="1"
      position={1}
      removeImage={[Function]}
      switchImages={[Function]}
      url="bar.jpg"
    />
    <div
      data-stylename="add-image"
    >
      <Loading />
    </div>
    <Connect(ChangeImageButton)
      attachmentType="image"
      update={[Function]}
      uploadSettings={
        Object {
          "id": 1,
          "type": "post",
        }
      }
    >
      <div
        data-stylename="add-image"
      >
        +
      </div>
    </Connect(ChangeImageButton)>
  </div>
</div>
`;

exports[`ImagePreview matches last snapshot 1`] = `
<div
  data-stylename="image-preview"
>
  <div
    data-stylename="image"
    style={
      Object {
        "backgroundImage": "url(foo.zng)",
      }
    }
  >
    <Icon
      data-stylename="remove-image"
      name="Ex"
      onClick={[Function]}
    />
    <div
      data-stylename="drag-preview"
    />
  </div>
</div>
`;
